tests/test_api.py

```diff
@@ -1,113 +1,174 @@
+        "distance_from_home": float(sample["distance_from_home"]),
+        "distance_from_last_transaction": float(sample["distance_from_last_transaction"]),
+        "ratio_to_median_purchase_price": float(sample["ratio_to_median_purchase_price"]),
+        "repeat_retailer": int(sample["repeat_retailer"]),
+        "used_chip": int(sample["used_chip"]),
+        "used_pin_number": int(sample["used_pin_number"]),
+        "online_order": int(sample["online_order"]),
+    }
+
+    for attempt in range(pred_data.max_retries):
+        try:
+            response = requests.post(CLIENT_URL, json=payload, timeout=500)
+            response.raise_for_status()
+            pred = response.json().get("prediction")
+            if pred is not None:
+                pred_data.predictions.append(pred)
+                pred_data.success = True
+                break
+        except RequestException as e:
+            pred_data.last_exception = e
+            print(f"Attempt {attempt + 1} failed for sample {idx}. Error: {str(e)}")
+            if attempt < pred_data.max_retries - 1:
+                time.sleep(pred_data.retry_delay)
+
+
+def test_api_accuracy(samples_data: Tuple) -> None:
     """
-    x_scaled, y_true = data_to_predict
-    predictions = []
-
-    for _, sample in enumerate(x_scaled):
-        payload = {
-            "distance_from_home": float(sample[0]),
-            "distance_from_last_transaction": float(sample[1]),
-            "ratio_to_median_purchase_price": float(sample[2]),
-            "repeat_retailer": int(sample[3]),
-            "used_chip": int(sample[4]),
-            "used_pin_number": int(sample[5]),
-            "online_order": int(sample[6]),
-        }
-
-        # Send POST request to the client API
-        response = requests.post(CLIENT_URL, json=payload, timeout=500)
-
-        # Verify the request was successful
-        assert (
-            response.status_code == 200
-        ), f"Request failed with status {response.status_code}"
-
-        # Retrieve the prediction
-        pred = response.json().get("prediction")
-        assert pred is not None, "Prediction is None"
-        predictions.append(pred)
-
-    # Calculate accuracy
-    accuracy = accuracy_score(y_true, predictions)
-    print(f"Accuracy on 10 predictions: {accuracy * 100:.2f}%")
-
-    # Verify the accuracy is acceptable (e.g., at least 70%)
-    assert accuracy >= 0.7, f"Accuracy too low: {accuracy * 100:.2f}%"
+    Sends 100 predictions via the client API and checks the accuracy.
+
+    Args:
+        samples_data: Tuple containing test samples and their true labels.
+    """
+    x_samples, y_true = samples_data
+    pred_data = PredictionData(
+        predictions=[],
+        max_retries=3,
+        retry_delay=5,
+        success=False
+    )
+
+    for i in range(30):
+        if check_server():
+            break
+        if i == 29:
+            pytest.fail("Server not accessible after 30 attempts")
+        time.sleep(1)
```

Comment:
isn't ths which is making the api slow?

Reply:
lt is a real possibility we have to fix that